### PR TITLE
Persist ext txn hashes

### PR DESCRIPTION
**Pull request by jagathweerasinghe-da**
[ci]

#### Cluster Testing
- [ ] If a cluster test is required, comment `/cluster_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a hard-migration test is required (from the latest release), comment `/hdm_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.

#### PR Guidelines
- [ ] Include any change that might be observable by our partners or affect their deployment in the [release notes](https://github.com/hyperledger-labs/splice/blob/main/docs/src/release_notes.rst).
- [ ] Specify fixed issues with `Fixes #n`, and mention issues worked on using `#n`
- [ ] Include a screenshot for frontend-related PRs - see [README](https://github.com/hyperledger-labs/splice/blob/main/TESTING.md#running-and-debugging-integration-tests) or use your favorite screenshot tool

#### Merge Guidelines
- [ ] Make the git commit message look sensible when squash-merging on GitHub (most likely: just copy your PR description).

### CODE REVIEW

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala [hunk 1]
```diff
@@ -957,5 +958,65 @@ class UpdateHistoryTest extends UpdateHistoryTestBase {
...
+        )
+      } yield {
+        updates should have size 1
+        externalTransactionHash should not be ByteString.EMPTY
+        val storedTransaction =
+          updates.loneElement.update.update.asInstanceOf[TransactionTreeUpdate].tree
+        storedTransaction.getExternalTransactionHash should be(externalTransactionHash)
+        storedTransaction.getExternalTransactionHash should be(
+          expectedUpdate.getExternalTransactionHash
+        )
```
[the remark this answers is not in the document]
Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala [hunk 2]
```diff
@@ -45,18 +48,22 @@ class TokenStandardCliIntegrationTest
       val (bobPublicKeyPath, bobPrivateKeyPath) =
         writeKeysToTempFile("bob", bobPublicKey, bobPrivateKey)
 
+      var extTxnHashes: Seq[String] = Seq.empty[String];
```
[the remark this answers is not in the document]
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala]: done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala [hunk 3]
```diff
@@ -142,12 +143,71 @@ trait UpdateHistoryTestUtil extends TestCommon {
...
+            .mkString
+        val recordedExtTxnHash =
+          Option(recordedTx.getExternalTransactionHash)
+            .map(_.toByteArray)
+            .getOrElse(Array.emptyByteArray)
+            .map("%02x" format _)
+            .mkString
+        actualExtTxnHash shouldBe recordedExtTxnHash
+      case _ =>
+        ()
```
[the remark this answers is not in the document]
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala [hunk 4]
```diff
@@ -179,20 +186,43 @@ class TokenStandardCliIntegrationTest
...
+      }
+
+      clue("SV1 scan backend should see the updates with external txn hashes") {
+        eventually() {
+          compareHistory(
+            sv1Backend.participantClient,
+            sv1ScanBackend.appState.automation.updateHistory,
+            sv1LedgerBeginOffset,
+            mustCheckExternalTxnHash = true,
+            extTxnHashes = extTxnHashes,
```
Comment (ray-roestenburg-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala]:
```suggestion
            extTxnHashes = Seq(onboardingAliceExternalTxnHash, onboardingAliceExternalTxnHash)
```
and remove the var in the beginning

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala [hunk 5]
```diff
@@ -142,12 +143,70 @@ trait UpdateHistoryTestUtil extends TestCommon {
...
+          .map(_.toByteArray)
+          .getOrElse(Array.emptyByteArray)
+          .map("%02x" format _)
+          .mkString
+    }
+
+    recordedExtTxnHashes should not be empty
+    recordedExtTxnHashes should contain allElementsOf extTxnHashes
+
+    actualUpdates.zip(recordedUpdates).foreach { case (actual, recorded) =>
```
Comment (ray-roestenburg-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: rather use forAll(...) { to collect all failures

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala [hunk 6]
```diff
@@ -142,12 +143,70 @@ trait UpdateHistoryTestUtil extends TestCommon {
...
+      compareExtTxnHashPerUpdate(actual, recorded)
+    }
+    succeed
+  }
+
+  private def compareExtTxnHashPerUpdate(
+      actual: UpdateHistoryResponse,
+      recorded: UpdateHistoryResponse,
+  ): Assertion =
+    (actual, recorded) match {
```
Comment (ray-roestenburg-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: you can do `val UpdateHistoryResponse(TransactionTreeUpdate(actualTx), _) = actual` and then use actualX (and then same for recordedTx)

File: apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala [hunk 7]
```diff
@@ -587,17 +587,19 @@ class UpdateHistory(
       .map(lengthLimited)
     val safeWorkflowId = lengthLimited(tree.getWorkflowId)
     val safeCommandId = lengthLimited(tree.getCommandId)
+    val safeExternalTransactionHash = tree.getExternalTransactionHash
```
Comment (ray-roestenburg-da) [inline apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala]: why is this 'safe'? (normally the result of a lengthLimited call)

Comment (ray-roestenburg-da) [reply apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala]: There is a lengthlimited bytestring that you can use. The hash should have a known length

File: apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala [hunk 8]
```diff
@@ -2435,6 +2444,7 @@ object UpdateHistory {
       rootEventIds: Seq[String],
       workflowId: Option[String],
       commandId: Option[String],
+      externalTransactionHash: Option[Array[Byte]],
```
Comment (ray-roestenburg-da) [inline apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala]: Looking at the code I'm not sure if you are consistent with Option[Array[Byte]] vs Array.empty vs null, maybe since this is marked as NonNull in javaapi.data, you could have this as a normal Array[Byte], so it's Array.empty or an actual hash

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala [hunk 9]
```diff
@@ -957,5 +958,65 @@ class UpdateHistoryTest extends UpdateHistoryTestBase {
...
+              externalTransactionHash = externalTransactionHash,
+            )
+          })(store)
+          updates <- store.getAllUpdates(
+            None,
+            PageLimit.Max,
+          )
+        } yield {
+          updates should have size 1
+          externalTransactionHash should be(ByteString.EMPTY)
```
Comment (ray-roestenburg-da) [inline apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: this is always true because of line 965

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala [hunk 10]
```diff
@@ -957,5 +958,65 @@ class UpdateHistoryTest extends UpdateHistoryTestBase {
...
+          val storedTransaction =
+            updates.loneElement.update.update.asInstanceOf[TransactionTreeUpdate].tree
```
Comment (ray-roestenburg-da) [inline apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: rather match instead of asInstanceOf

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala [hunk 11]
```diff
@@ -957,5 +958,65 @@ class UpdateHistoryTest extends UpdateHistoryTestBase {
...
+              externalTransactionHash = externalTransactionHash,
+            )
+          })(store)
+          updates <- store.getAllUpdates(
+            None,
+            PageLimit.Max,
+          )
+        } yield {
+          updates should have size 1
+          externalTransactionHash should not be ByteString.EMPTY
```
Comment (ray-roestenburg-da) [inline apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: always true since externalTransactionHash is a val, you probably meant something else

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala [hunk 12]
```diff
@@ -957,5 +958,65 @@ class UpdateHistoryTest extends UpdateHistoryTestBase {
...
+          val storedTransaction =
+            updates.loneElement.update.update.asInstanceOf[TransactionTreeUpdate].tree
```
Comment (ray-roestenburg-da) [inline apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: match instead

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTestBase.scala [hunk 13]
```diff
@@ -372,7 +372,7 @@ object UpdateHistoryTestBase {
       /*traceContext = */ TraceContextOuterClass.TraceContext.getDefaultInstance, // Not preserved
 
       /*recordTime = */ tree.getRecordTime,
-      /*externalTransactionHash = */ ByteString.EMPTY,
+      /*externalTransactionHash = */ tree.getExternalTransactionHash,
```
Comment (ray-roestenburg-da) [inline apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTestBase.scala:375]: Ensure that you always use array empty or an array and avoid null (I haven't fully checked but this thought come to mind)

Comment (ray-roestenburg-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTestBase.scala:375]: (Though on second thought in postgres NULL makes more sense especially with possibly existing data, so you probably did the right thing here)

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/ExternallySignedPartyTestUtil.scala [hunk 14]
```diff
@@ -187,15 +190,19 @@ trait ExternallySignedPartyTestUtil extends TestCommon {
       externalPartyOnboarding: OnboardingResult,
       proposal: ExternalPartySetupProposal.ContractId,
       verboseHashing: Boolean = false,
-  )(implicit env: SpliceTestConsoleEnvironment): (TransferPreapproval.ContractId, String) = {
+  )(implicit
+      env: SpliceTestConsoleEnvironment
+  ): (TransferPreapproval.ContractId, String, String) = {
```
Comment (ray-roestenburg-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/ExternallySignedPartyTestUtil.scala]: Prefer a case class with fields so you don't have to think about which String is which, and easier to get the fields out where you use it.

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala [hunk 15]
```diff
@@ -45,18 +48,22 @@ class TokenStandardCliIntegrationTest
       val (bobPublicKeyPath, bobPrivateKeyPath) =
         writeKeysToTempFile("bob", bobPublicKey, bobPrivateKey)
 
+      var extTxnHashes: Seq[String] = Seq.empty[String]
```
Comment (ray-roestenburg-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala]: you don't need this var

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala [hunk 16]
```diff
@@ -105,10 +104,12 @@ trait UpdateHistoryTestUtil extends TestCommon {
       updateHistory: UpdateHistory,
       ledgerBegin: Long,
       mustIncludeReassignments: Boolean = false,
+      mustCheckExternalTxnHash: Boolean = false,
```
Comment (ray-roestenburg-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: In general rather have better named methods than flags to a method, so add a def compareHistoryIncludingHashes or something like that

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala (hunk 9 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: gone

File: apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala (hunk 7 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala]: done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala (hunk 5 above)
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: done

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala (hunk 11 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: gone

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala (hunk 10 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala (hunk 4 above)
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala]: done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala (hunk 15 above)
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TokenStandardCliIntegrationTest.scala]: gone

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala (hunk 16 above)
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: done

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala (hunk 12 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/ExternallySignedPartyTestUtil.scala (hunk 14 above)
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/ExternallySignedPartyTestUtil.scala]: done

File: apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala (hunk 8 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala]: (1) @NonNull doesn't guarantee it is always not null, it is just an indicator
(2) Even if we get a non-null value from the Ledger APIs, scan db can still have null for the historical data. Am I missing something there?
(3) Option[Array[Byte]] is exactly the same approach taken in ledger-side for the DTO.

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala (hunk 6 above)
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: We can't just enforce it that way when we could have ReassignmentUpdate for example, right?

Comment (ray-roestenburg-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: ah right my bad, that's correct

Comment (ray-roestenburg-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: I guess more natural for me would have been something like 
(actual.update, recorded.update) match { case TransactionTreeUpdate(actualTx) => ...

Since both are of type UpdateHistoryResponse

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala [hunk 17]
```diff
@@ -142,12 +168,64 @@ trait UpdateHistoryTestUtil extends TestCommon {
...
+      actual: UpdateHistoryResponse,
+      recorded: UpdateHistoryResponse,
+  ): Assertion =
+    (actual, recorded) match {
+      case (
+            UpdateHistoryResponse(TransactionTreeUpdate(actualTx), _),
+            UpdateHistoryResponse(TransactionTreeUpdate(recordedTx), _),
+          ) =>
+        val actualExtTxnHash =
+          Option(actualTx.getExternalTransactionHash)
```
Comment (ray-roestenburg-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: nitpick you could do this in a inline def since it's repeated

File: apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala (hunk 8 above)
Comment (ray-roestenburg-da) [reply apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala]: ok, thanks. I guess I would have liked it better if we only had to deal with array empty or an array, and not the none case as well, which just means the same thing, and I'm wondering if you can map it as such in slick

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/StoreTestBase.scala [hunk 18]
```diff
@@ -1314,7 +1315,7 @@ abstract class StoreTestBase
       synchronizerId.toProtoPrimitive,
       TraceContextOuterClass.TraceContext.getDefaultInstance,
       recordTime,
-      ByteString.EMPTY,
+      externalTransactionHash,
```
Comment (ray-roestenburg-da) [inline apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/StoreTestBase.scala:1318]: Do you also explicitly test with null somewhere?

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala (hunk 12 above)
Comment (ray-roestenburg-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: I don't see it, the asInstanceOf is still there

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala (hunk 17 above)
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: done

File: apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala (hunk 8 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/main/scala/org/lfdecentralizedtrust/splice/store/UpdateHistory.scala]: mapping in slick done.

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala (hunk 12 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: ah missed this one. introduced a def.

Comment (ray-roestenburg-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: I think we are misunderstanding eachother @jagathweerasinghe-da , I mean:

```
updates.loneElement.update.update.asInstanceOf[TransactionTreeUpdate].tree
```
On line 1013.
Change this to something like 
```
val storedTransaction = updates.loneElement.update.update match {
  case TransactionTreeUpdate(tree) => tree
  case u => fail(s"unexpected update $u")
}
```

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/StoreTestBase.scala (hunk 18 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/StoreTestBase.scala:1318]: added.

File: apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala (hunk 12 above)
Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: done

Comment (jagathweerasinghe-da) [reply apps/common/src/test/scala/org/lfdecentralizedtrust/splice/store/UpdateHistoryTest.scala]: used in a def as it is needed multiple times.

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala (hunk 6 above)
Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: nice, done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala [hunk 19]
```diff
@@ -142,12 +169,59 @@ trait UpdateHistoryTestUtil extends TestCommon {
       actualUpdates.map(UpdateHistoryTestBase.withoutLostData(_, mode = LostInStoreIngestion))
     val recordedUpdatesWithoutLostData = recordedUpdates.map(_.update)
     actualUpdatesWithoutLostData should have length recordedUpdatesWithoutLostData.size.longValue()
+
     actualUpdatesWithoutLostData.zip(recordedUpdatesWithoutLostData).foreach {
```
Comment (ray-roestenburg-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: this probably can also use forAll() from scalatest, you probably get better feedback when it fails.

Comment (jagathweerasinghe-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/util/UpdateHistoryTestUtil.scala]: done
